### PR TITLE
Feature/profile

### PR DESCRIPTION
**Pull request by Miscellian**

### CODE REVIEW

File: GameSalesApi/GameSalesApi/Features/AccountManagement/AccountController.cs [hunk 1]
```diff
@@ -1,32 +1,38 @@
...
+using Microsoft.AspNetCore.Mvc;
+using Microsoft.Extensions.Logging;
 using Model;
+using System.Collections.Generic;
 
 namespace GameSalesApi.Features.AccountManagement
 {
     [ApiController]
     [Route("api/[controller]")]
+
```
Comment (AnatoIii) [inline GameSalesApi/GameSalesApi/Features/AccountManagement/AccountController.cs]: ?

File: GameSalesApi/GameSalesApi/Features/AccountManagement/CommandHandlers/UpdateUserCommandHandler.cs [hunk 2]
```diff
@@ -66,11 +74,12 @@ public override Result Handle(UpdateUserCommand input)
...
                 LastName = input.LastName ?? user.LastName,
                 Email = input.Email ?? user.Email,
+                Username = input.Username ?? user.Username,
                 NotificationViaEmail = input.NotificationViaEmail ?? user.NotificationViaEmail,
                 NotificationViaTelegram = input.NotificationViaTelegram ?? user.NotificationViaTelegram,
-                PhotoLink = input.PhotoLink ?? user.PhotoLink,
                 PasswordSalt = passwordSalt != null ? Convert.ToBase64String(passwordSalt) : user.PasswordSalt,
-                PasswordHash = passwordHash ?? user.PasswordHash
+                PasswordHash = passwordHash ?? user.PasswordHash,
+                PhotoLink = user.PhotoLink
```
Comment (AnatoIii) [inline GameSalesApi/GameSalesApi/Features/AccountManagement/CommandHandlers/UpdateUserCommandHandler.cs]: ?? Why you update photolink in user edit method?

File: GameSalesApi/GameSalesApi/Features/AccountManagement/CommandHandlers/UploadProfilePhotoCommandHandler.cs [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+                NotificationViaEmail = user.NotificationViaEmail,
+                NotificationViaTelegram = user.NotificationViaTelegram,
+                PhotoLink = photoLink ?? user.PhotoLink,
+                PasswordSalt = user.PasswordSalt,
+                PasswordHash = user.PasswordHash
+            };
+
+            user.UpdateUser(newUser);
+
+            _rDBContext.Users.Update(user);
```
Comment (AnatoIii) [inline GameSalesApi/GameSalesApi/Features/AccountManagement/CommandHandlers/UploadProfilePhotoCommandHandler.cs]: Why not to update user.PhotoLink?

File: GameSalesApi/GameSalesApi/Features/AccountManagement/Commands/UpdateUserCommand.cs [hunk 4]
```diff
@@ -21,6 +21,16 @@ public class UpdateUserCommand : ICommand<Result>
         /// </summary>
         public string Email { get; set; }
 
+        /// <summary>
+        /// New username
```
Comment (AnatoIii) [inline GameSalesApi/GameSalesApi/Features/AccountManagement/Commands/UpdateUserCommand.cs:25]: New?

File: GameSalesApi/GameSalesApi/Features/AccountManagement/ImageService.cs [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+            _imgurConfig = options.Value;
+            _client = new HttpClient();
+        }
+
+        public async Task<IImage> UploadImage(IFormFile image)
+        {
+            var client = new ImgurClient(_imgurConfig.ClientId);
+            var endpoint = new ImageEndpoint(client);
+            IImage uploadedImage = await endpoint.UploadImageStreamAsync(image.OpenReadStream());
+            return uploadedImage;
```
Comment (AnatoIii) [inline GameSalesApi/GameSalesApi/Features/AccountManagement/ImageService.cs]: return endpoint...
And we don't use services

File: GameSalesApi/GameSalesApi/Features/Authorization/TokenCreator.cs [hunk 6]
```diff
@@ -78,7 +78,7 @@ public TokenDTO CreateDTOToken(User user, GameSalesContext dbContext)
                 DueDate = DateTime.Now.AddMinutes(_tokenConfig.RefreshTokenLifetime)
             };
             dbContext.Tokens.Add(dbToken);
-            return new TokenDTO() { AccessToken = accessToken, RefreshToken = refreshTokenValue };
+            return new TokenDTO() { AccessToken = accessToken, RefreshToken = refreshTokenValue, UserId = user.Id};
```
Comment (AnatoIii) [inline GameSalesApi/GameSalesApi/Features/Authorization/TokenCreator.cs]: You can get user.Id from token, just parse it on frontend

File: GameSalesApi/GameSalesApi/Features/Authorization/TokenDTO.cs [hunk 7]
```diff
@@ -12,5 +9,6 @@ public class TokenDTO
     {
         public string RefreshToken { get; set; }
         public string AccessToken { get; set; }
+        public Guid UserId { get; set; }
```
Comment (AnatoIii) [inline GameSalesApi/GameSalesApi/Features/Authorization/TokenDTO.cs]: The same

File: GameSalesApi/GameSalesApi/appsettings.json [hunk 8]
```diff
@@ -7,13 +7,17 @@
     }
   },
   "ConnectionStrings": {
-    "GameSalesContext": ""
+    "GameSalesContextSecret": "Host=77.47.208.42;Port=6500;Database=GameSales;Username=salesservice;Password=salespassword"
```
Comment (AnatoIii) [inline GameSalesApi/GameSalesApi/appsettings.json]: ...

File: GameSalesApi/GameSalesApi/appsettings.json [hunk 9]
```diff
@@ -7,13 +7,17 @@
...
     "JWTLifetime": 60,
     "RefreshTokenLifetime": 43200,
     "Secret": "SuperMegaSecretSigningKeyThatIsActuallyImpossibleToGuess",
     "Issuer": "https://localhost:5000",
     "Audience": "GamesSales"
+  },
+  "ImgurApi": {
+    "ClientID": "a54fb71361d291e",
+    "ClientSecret": "82a1fffdb29ca6af540fc9dc979bde61a5a2faba"
   }
```
Comment (AnatoIii) [inline GameSalesApi/GameSalesApi/appsettings.json]: Yeah, its great idea to commit secret data)

File: game-sales-tracker/src/app/app-routing/app-routing.module.ts [hunk 10]
```diff
@@ -1,22 +1,28 @@
...
-  imports: [CommonModule, RouterModule.forRoot(routes)],
-  exports: [RouterModule],
-  declarations: [],
+    imports: [CommonModule, RouterModule.forRoot(routes)],
+    exports: [RouterModule],
+    declarations: [],
 })
-export class AppRoutingModule {}
+export class AppRoutingModule {
+}
```
Comment (AnatoIii) [inline game-sales-tracker/src/app/app-routing/app-routing.module.ts]: One line

File: game-sales-tracker/src/app/app.module.ts [hunk 11]
```diff
@@ -1,41 +1,58 @@
...
+            multi: true,
+        },
+    ],
+    bootstrap: [AppComponent],
+    entryComponents: [SignInComponent],
+    exports: [HeaderComponent],
 })
-export class AppModule {}
+export class AppModule {
+}
```
Comment (AnatoIii) [inline game-sales-tracker/src/app/app.module.ts]: One line

File: game-sales-tracker/src/app/models/token-dto.ts [hunk 12]
```diff
@@ -1,4 +1,5 @@
 export interface TokenDto {
     accessToken: string;
     refreshToken: string;
+    userId: string;
```
Comment (AnatoIii) [inline game-sales-tracker/src/app/models/token-dto.ts]: From token

File: game-sales-tracker/src/app/models/user.ts [hunk 13]
```diff
@@ -0,0 +1,9 @@
+export interface User {
```
Comment (AnatoIii) [inline game-sales-tracker/src/app/models/user.ts]: Start interfaces from I -> IUser

File: game-sales-tracker/src/app/profile/profile.component.html [hunk 14]
```diff
@@ -0,0 +1,182 @@
...
...oeorTYAzZkwwoMz19pzj9bnH/GP/+qbchjSGflneWYhtTuKdMOmNKZcJ5TjInQKcYXnESd/jQxy0ENpULTNGOGgxpap/oyw9pbUAqhfx2Dbkhovvfgz4iUzoM9+GlK6/Mh4q29hyC1mwro30hpVVLPF9wYQr71RazOeM5/cw81iBRD+A03aM9/C/obbrKjbYSpCmIVG3qT/Q8oeUo3Rz0IL7vI1tEbCB9pSiu8I/aV8x3Kg/BGWrWp4ZVs0nZfmAoEG4h/61yHYIJiFSl6Q0Vk6tTW1N8kYp8hdOkfHYYMXd2Qft+8CYwqYDSKvqIh+MCF8Wgca2u/cwdgeW3TtuVn6+1oBs3yLo5C2JpK6CvQzGpfUkz9UG/87gCsi5o2LIXolxN0FbwAsjOLEr+YJmXn7iR6N0BCt5p5cMxm7eAsfS+/CACQf4CTpKjzgkvr2cVarVTf96372yut7XLJ1sa7lv6VcfgYrWaxqr3Wlo1S6pvStr22sxOtTNPLzdY3nj20bPP+ejFdJYkLsjGLdtPBEbe/mr2bQKiXWJDroA+vtzc0p9aahuwqHMDYrQEXHEw9jwQl3drMpts9JBU1SdktPe5FBRdJQ6bwXBpa57ib2A8kukQDzMjh++Uo7Fo6Wd02Pkf4fknqoo4HtvAIjsqUcjx6DIPgWCaOML9rKI/oqD9/lgNrn+eF+p7j8tnzHBiR7+kdUGw/+V1Kzkc75mMy6U+FMaxjPibiM1U1uGM+puInHpmALZCgP4pt7i840MV8+0R1zPsRB6UTcqpizncYwZ89syDydfyWCwXB1l8/zRNGWbTG/GHKUm9AkxHMc/EGSk3z2+ArEhPEV5TUBLEvUGFcjEUH80J/jveTGOAJEljJbILWGQT3zRYiwuKsUXN1EEJAzBhRJFll7mBUG7KD8EqPkKekBREaL8hMDZLQSG6AQjtHPYmvTQnX0TtpC1SYCe2YdkkyLP3jj5BSbKiuR585eQhTgoje6yIb0Yb0C+mV6EYvebqw5SDy2WmubogZiF2AVxPC2FpDf8H2Q9QWo6IkjUxTWVEI3WY/wrCeSuqJ+eRWzXR/JXwgVjUMozbCOfoEZiSiKVGepqv5CJ8RyR4D7xBeamqa7z3BJ/z17JxuBPdv93d/a2Ki878MMAzDMAzDMAzDMAzDMF/KP09VUmxBAiI3AAAAAElFTkSuQmCC'"
```
Comment (AnatoIii) [inline game-sales-tracker/src/app/profile/profile.component.html]: Amd what is it?

**Review by AnatoIii — CHANGES_REQUESTED** [review]

File: GameSalesApi/GameSalesApi/Features/AccountManagement/AccountController.cs [hunk 15]
```diff
@@ -1,32 +1,38 @@
...
 namespace GameSalesApi.Features.AccountManagement
 {
     [ApiController]
     [Route("api/[controller]")]
+ 
     public class AccountController : ControllerBase
     {
         private readonly GameSalesContext _rDBContext;
+
+        private readonly ImageService _rImageService;
```
Comment (JokerFunny) [inline GameSalesApi/GameSalesApi/Features/AccountManagement/AccountController.cs]: read-only field could be together without newline(they are logically combined)

File: GameSalesApi/GameSalesApi/Features/AccountManagement/CommandHandlers/UpdateUserCommandHandler.cs [hunk 16]
```diff
@@ -51,6 +51,14 @@ public override Result Handle(UpdateUserCommand input)
             byte[] passwordSalt = null;
             string passwordHash = null;
 
+            if (input.Password != null && !PasswordHelpers.ValidatePassword(input.CurrentPassword, user.PasswordSalt, user.PasswordHash))
+                return Result.Fail("Incorrect password");
+
+            if (input.Password == input.CurrentPassword && input.CurrentPassword != null)
+            {
+                return Result.Fail("Current and new passwords are same");
+            }
```
Comment (JokerFunny) [inline GameSalesApi/GameSalesApi/Features/AccountManagement/CommandHandlers/UpdateUserCommandHandler.cs]: Remove redundant braces {}

File: GameSalesApi/GameSalesApi/Features/AccountManagement/CommandHandlers/UpdateUserCommandHandler.cs [hunk 17]
```diff
@@ -51,6 +51,14 @@ public override Result Handle(UpdateUserCommand input)
             byte[] passwordSalt = null;
             string passwordHash = null;
 
+            if (input.Password != null && !PasswordHelpers.ValidatePassword(input.CurrentPassword, user.PasswordSalt, user.PasswordHash))
```
Comment (JokerFunny) [inline GameSalesApi/GameSalesApi/Features/AccountManagement/CommandHandlers/UpdateUserCommandHandler.cs]: split with new line

File: GameSalesApi/GameSalesApi/Features/AccountManagement/CommandHandlers/UploadProfilePhotoCommandHandler.cs [hunk 18]
```diff
@@ -0,0 +1,80 @@
...
+using GameSalesApi.Features.AccountManagement.Commands;
+using GameSalesApi.Helpers;
+using Infrastructure.HandlerBase;
+using Infrastructure.Result;
+using Model;
+using System;
+
+namespace GameSalesApi.Features.AccountManagement.CommandHandlers
+{
+    public class UploadProfilePhotoCommandHandler
```
Comment (JokerFunny) [inline GameSalesApi/GameSalesApi/Features/AccountManagement/CommandHandlers/UploadProfilePhotoCommandHandler.cs:16]: All public class, methods, props and ctors should have summary

File: GameSalesApi/GameSalesApi/Features/AccountManagement/CommandHandlers/UploadProfilePhotoCommandHandler.cs [hunk 19]
```diff
@@ -0,0 +1,80 @@
...
+        /// <param name="imageService"><see cref="ImageService"/></param>
+        public UploadProfilePhotoCommandHandler(GameSalesContext dbContext, ImageService imageService)
+            : base(null)
+        {
+            _rDBContext = dbContext;
+            _imageService = imageService;
+        }
+
+        /// <summary>
+        /// Update <see cref="User"/> in <see cref="GameSalesContext"/>
```
Comment (JokerFunny) [inline GameSalesApi/GameSalesApi/Features/AccountManagement/CommandHandlers/UploadProfilePhotoCommandHandler.cs]: Specify that updates User.PhotoLink

File: GameSalesApi/GameSalesApi/Features/AccountManagement/CommandHandlers/UploadProfilePhotoCommandHandler.cs [hunk 20]
```diff
@@ -0,0 +1,80 @@
...
+        /// </summary>
+        /// <param name="command"><see cref="UploadProfilePhotoCommand"/></param>
+        public override void Execute(UploadProfilePhotoCommand command)
+            => Handle(command);
+
+        /// <summary>
+        /// Update <see cref="User"/> in <see cref="GameSalesContext"/>
```
Comment (JokerFunny) [inline GameSalesApi/GameSalesApi/Features/AccountManagement/CommandHandlers/UploadProfilePhotoCommandHandler.cs]: Same

File: GameSalesApi/GameSalesApi/Features/AccountManagement/Commands/UpdateUserCommand.cs (hunk 4 above)
Comment (JokerFunny) [reply GameSalesApi/GameSalesApi/Features/AccountManagement/Commands/UpdateUserCommand.cs:25]: We can't change username or what question?

File: GameSalesApi/GameSalesApi/Features/AccountManagement/Commands/UploadProfilePhotoCommand.cs [hunk 21]
```diff
@@ -0,0 +1,23 @@
+﻿using Infrastructure.CommandBase;
+using Infrastructure.Result;
+using Microsoft.AspNetCore.Http;
+using System;
+
+
```
Comment (JokerFunny) [inline GameSalesApi/GameSalesApi/Features/AccountManagement/Commands/UploadProfilePhotoCommand.cs:6]: ??

File: GameSalesApi/GameSalesApi/Features/AccountManagement/Commands/UploadProfilePhotoCommand.cs [hunk 22]
```diff
@@ -0,0 +1,23 @@
...
+    public class UploadProfilePhotoCommand: ICommand<Result>
+    {
+        /// <summary>
+        /// User profile photo
+        /// </summary>
+        public IFormFile Image { get; set; }
+        /// <summary>
+        /// User Id
+        /// </summary>
+        public Guid UserId { get; set; }
```
Comment (JokerFunny) [inline GameSalesApi/GameSalesApi/Features/AccountManagement/Commands/UploadProfilePhotoCommand.cs:17]: Place it like a first param

File: GameSalesApi/GameSalesApi/Features/AccountManagement/ImageService.cs [hunk 23]
```diff
@@ -0,0 +1,33 @@
...
+namespace GameSalesApi.Features.AccountManagement
+{
+    /// <summary>
+    /// A service, responsible for managing user images
+    /// </summary>
+    public class ImageService
+    {
+        private readonly ImgurConfig _imgurConfig;
+        private readonly HttpClient _client;
+        public ImageService(IOptions<ImgurConfig> options)
```
Comment (JokerFunny) [inline GameSalesApi/GameSalesApi/Features/AccountManagement/ImageService.cs]: summary and new line

File: GameSalesApi/GameSalesApi/Features/AccountManagement/ImageService.cs [hunk 24]
```diff
@@ -0,0 +1,33 @@
...
+        {
+            _imgurConfig = options.Value;
+            _client = new HttpClient();
+        }
+
+        public async Task<IImage> UploadImage(IFormFile image)
```
Comment (JokerFunny) [inline GameSalesApi/GameSalesApi/Features/AccountManagement/ImageService.cs]: summary

File: GameSalesApi/GameSalesApi/Features/AccountManagement/ImageService.cs [hunk 25]
```diff
@@ -0,0 +1,33 @@
...
+using System;
+using System.Net.Http;
+using System.Threading.Tasks;
+
+namespace GameSalesApi.Features.AccountManagement
+{
+    /// <summary>
+    /// A service, responsible for managing user images
+    /// </summary>
+    public class ImageService
```
Comment (JokerFunny) [inline GameSalesApi/GameSalesApi/Features/AccountManagement/ImageService.cs]: Better rework it on command if you want to use it on POST/DELETE/UPDATE or query if you want to use it with GET

File: GameSalesApi/GameSalesApi/ImgurConfig.cs [hunk 26]
```diff
@@ -0,0 +1,8 @@
+﻿namespace GameSalesApi
+{
+    public class ImgurConfig
```
Comment (JokerFunny) [inline GameSalesApi/GameSalesApi/ImgurConfig.cs:7]: summary

File: GameSalesApi/GameSalesApi/ImgurConfig.cs [hunk 27]
```diff
@@ -0,0 +1,8 @@
...
+    {
+        public string ClientId { get; set; }
+        public string ClientSecret { get; set; }
```
Comment (JokerFunny) [inline GameSalesApi/GameSalesApi/ImgurConfig.cs]: same

**Review by JokerFunny — CHANGES_REQUESTED** [review]

**Review by AnatoIii — APPROVED** [review]

**Pull request merged**